pkg/registry/flowcontrol/rest/flowschema.go

```diff
@@ -0,0 +1,149 @@
+	// It returns true if auto update is required and the returned object will represent the new configuration to be stored in-cluster.
+	ShouldUpdate(current, bootstrap *flowcontrolv1beta1.FlowSchema) (*flowcontrolv1beta1.FlowSchema, bool)
+}
+
+func ensureFlowSchema(clientset flowcontrolclient.FlowcontrolV1beta1Interface, fsType string, bootstrap *flowcontrolv1beta1.FlowSchema, strategy fsEnsureStrategy) error {
+	current, err := clientset.FlowSchemas().Get(context.TODO(), bootstrap.Name, metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to retrieve FlowSchema type=%s name=%q error=%w", fsType, bootstrap.Name, err)
+		}
+
+		switch strategy.ShouldCreate() {
+		case true:
+			if _, err := clientset.FlowSchemas().Create(context.TODO(), bootstrap, metav1.CreateOptions{FieldManager: thisFieldManager}); err != nil {
+				return fmt.Errorf("cannot create FlowSchema type=%s name=%q error=%w", fsType, bootstrap.Name, err)
+			}
+			klog.V(3).Infof("Successfully created FlowSchema type=%s name=%s", fsType, bootstrap.Name)
+		default:
+			klog.V(3).Infof("Skipping creation of FlowSchema type=%s name=%s", fsType, bootstrap.Name)
+		}
+		return nil
+	}
+
+	klog.V(3).Infof("The FlowSchema already exists, ensuring it is up to date type=%s name=%s", fsType, bootstrap.Name)
+	newObject, update := strategy.ShouldUpdate(current, bootstrap)
+	if !update {
+		klog.V(3).Infof("No update required for the FlowSchema type=%s name=%q diff: %s", fsType, bootstrap.Name, cmp.Diff(current, bootstrap))
+		return nil
+	}
+
+	if _, err := clientset.FlowSchemas().Update(context.TODO(), newObject, metav1.UpdateOptions{FieldManager: thisFieldManager}); err != nil {
+		return fmt.Errorf("failed to upgrade the FlowSchema, will retry later type=%s name=%q error=%w", fsType, bootstrap.Name, err)
+	}
+
+	klog.V(3).Infof("Updated the FlowSchema type=%s name=%q diff=%s", fsType, bootstrap.Name, cmp.Diff(current, newObject))
+	return nil
+}
+
+// auto-update strategy for the suggested FlowSchema objects
+type suggestedFSEnsureStrategy struct {
+	shouldCreate bool
+}
+
+func (s *suggestedFSEnsureStrategy) ShouldCreate() bool {
+	return s.shouldCreate
+}
+
+func (s *suggestedFSEnsureStrategy) ShouldUpdate(current, bootstrap *flowcontrolv1beta1.FlowSchema) (*flowcontrolv1beta1.FlowSchema, bool) {
+	if current == nil || bootstrap == nil {
+		return nil, false
+	}
+
+	updateSpec := shouldUpdateSpec(current)
+	updateAnnotationKey := shouldUpdateAnnotation(current, updateSpec)
+
+	var specChanged bool
+	switch {
+	case !updateSpec && !updateAnnotationKey:
+		// the annotation key is up to date and auto-update is disabled, no update is necessary
+		return nil, false
+	case updateSpec:
+		specChanged = flowSchemaSpecChanged(bootstrap, current)
+		if !updateAnnotationKey && !specChanged {
+			// the annotation key is up to date and the spec has not changed, no update is necessary
+			return nil, false
+		}
+	}
```

Comment:
I simplified the block, but I still need to retain `autoUpdateSpec` (previously `updateSpec`) because its value is what gets stored in the annotation.